[jit] bake constants into the traced graph, get rid of getNestedValueTrace

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21046 [jit] bake constants into the traced graph, get rid of getNestedValueTrace**
* #21003 [jit] remove constant pooling expect
* #21002 [jit] move casting ops from prim to aten

When tracing a script function, we will stop the tracing state and try to inline the scripted graph, but if the scripted function contains non tensor types, we will fail to trace them as they are not generated by any previous ops or any constants.

This PR helps generate the constants when the tracing call try to inline the script function. 

This should fix https://github.com/pytorch/pytorch/issues/21009

Differential Revision: [D15530991](https://our.internmc.facebook.com/intern/diff/D15530991)

wanchaol committed:
[jit] bake constants into the traced graph, get rid of getNestedValueTrace

wanchaol committed:
Update on "[jit] bake constants into the traced graph, get rid of getNestedValueTrace"

[jit] bake constants into the traced graph, get rid of getNestedValueTrace

gh-metadata: pytorch pytorch 21046 gh/wanchaol/13/head

wanchaol committed:
resolve conflict on "[jit] bake constants into the traced graph, get rid of getNestedValueTrace"

[jit] bake constants into the traced graph, get rid of getNestedValueTrace

gh-metadata: pytorch pytorch 21046 gh/wanchaol/13/head

eellison commented:
Why does is this stacked with the other PRs?

wanchaol commented:
> Why does is this stacked with the other PRs?

@eellison it has no relation with the bottom PRs, it's just that I am lazy on creating a new branch for it lol... I can move it to a separate branch if you prefer :)

zdevito commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -134,33 +149,12 @@ Value* getValueTrace(const IValue& var) {
-            v.toTuple()->elements(),
-            [](const IValue& val) { return getNestedValueTrace(val); })))
-        ->output();
+    // If the values are non-tensors, we try to create constants 
```
insertConstant of arbitrary IValues can fail with an internal assert. This needs to only be called if we _can_ create a constant from var.

wanchaol replied:
I checked that `graph->insertConstant` already call `tryInsertConstant` and do the error handling after https://github.com/pytorch/pytorch/pull/19686, it will throw constant_not_support error if the ivalues are not valid. Shall I simply add a test for it and we are good to go?

zdevito reviewed (commented):
Looks good, but needs to have error handling and a test for the error handline.

wanchaol committed:
Update on "[jit] bake constants into the traced graph, get rid of getNestedValueTrace"

[jit] bake constants into the traced graph, get rid of getNestedValueTrace

gh-metadata: pytorch pytorch 21046 gh/wanchaol/13/head

zdevito commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -151,9 +151,14 @@ Value* getValueTrace(const IValue& var) {
+      return *constant;
+    }
+    std::ostringstream os;
+    os << "Unknown type used in value trace lookup!";
```
Can this error do a better job of reporting what value is not supported as a constant? At the very least it should print the value it is trying to make a constant of.

wanchaol committed:
address comment on "[jit] bake constants into the traced graph, get rid of getNestedValueTrace"

[jit] bake constants into the traced graph, get rid of getNestedValueTrace

gh-metadata: pytorch pytorch 21046 gh/wanchaol/13/head

wanchaol committed:
Update on "[jit] bake constants into the traced graph, get rid of getNestedValueTrace"

[jit] bake constants into the traced graph, get rid of getNestedValueTrace

gh-metadata: pytorch pytorch 21046 gh/wanchaol/13/head
